### ISSUE TITLE
Repository structure revision: sdk/[x] instead of sdk/[service]

### ISSUE THREAD

**Issue by mitchdenny** [Azure/azure-sdk#892]
In our recent conversations with the Azure Spring Boot team we've been looking at what it might look like to pull the Azure Spring Boot modules (into the ```azure-sdk-for-java``` repo). The Azure Spring Boot modules cover a cross section of existing service areas as well as some that don't fit neatly into the service folders we have today.

I would like to propose that we loosen the definition of what a folder means under the ```sdk`` folder to include supporting libraries for particular frameworks such as Spring Boot. Whilst this issue came up in the context of the Java mono-repo I think it is worth getting a consensus on this for all our repos so when we have similar situations come up in other repos we don't have to debate the pros and cons again.

Applying this to the spring boot libraries, we would end up with this structure:

```
sdk/
  springboot/
    azure-spring-boot-bom
    azure-spring-boot-parent
    azure-spring-boot
    azure-servicebus-jms-spring-boot-starter
    ...
```

Alternatively, if there was a clear connection between a service directory and the spring boot library we could go with this structure:

```
sdk/
  springboot/
    azure-spring-boot-bom
    azure-spring-boot-parent
    azure-spring-boot
  servicebus/
    azure-servicebus-jms-spring-boot-starter
    ...
```

I prefer the first option because its more likely that the spring boot libraries are going to co-evolve together. Building them together will result in less developer friction - and they are more likely to need to ship together when something like the version of spring boot they are built against changes.

**Comment by mitchdenny** [comment]
Adding @weshaggard @JonathanGiles @joshfree @hemanttanwar @kurtzeborn @JimSuplizio to the issue so we can discuss and settle this when folks start filtering back from the holiday break.

**Comment by weshaggard** [comment]
@mitchdenny is the springboot any more then a set of dependency libraries? For example does it also include dramatically different toolsets? If it is just another dependency I think I would prefer that we go with your second option and have the libraries still tied to the service as opposed to the framework. I see this no different from the azure-core dependency. In fact it would be ideal if we could hide most of the springboot dependency in some flavor of an azure-core library. 

As for loosening the meaning of "service" I'm not too concerned with that as the most important thing for me is that we keep the structure consistent as "sdk/<something>/<package>" and hopefully the "<something>" ties together the set of packages in some way. If we felt that springboot was the largest tying factor then I'd be fine with naming that but at this point I still think the service calls are probably the largest tying factor so I'd lean towards keeping your second suggestion.

**Comment by Petermarcu** [comment]
+1 on service libraries ending up in the service folders. Core libraries for Springboot could follow Azure Core pattern.

**Comment by hemanttanwar** [comment]
I would agree with @weshaggard for structure.
I like to clarify few points.

1. Most of springboot modules will reside in sdk/springboot  : This will allow developer/users of Spring Boot community to point to one place where  everything related to "springboot " reside.

2. From Engineering toolset point of view, its no different and it can reuse existing tools.

3. We will try to keep Azure.core  framework agnostic as much as possible. 
   Azure.core is working at protocol layer like HTTP and I with to keep it that way.

    sdk/springboot will have its own equavalent of azure-core but more specific to springboot

4. Case by case basis we can discuss which sprongboot library need Azure.core like pattern.  
   
4. Most of the springboot could work with `sdk/<something>/<package` but following two might have one extra level and  we might have to restructure these 

a.  `sdk/springboot/spring-integration-azure/spring-integration-servicebus`  @ https://github.com/microsoft/spring-cloud-azure/tree/master/spring-integration-azure/spring-integration-servicebus

b. 'spring-cloud-azure-servicebus-stream-binder-core' @ https://github.com/microsoft/spring-cloud-azure/tree/master/spring-cloud-azure-stream-binder/spring-cloud-azure-servicebus-stream-binder-core

**Comment by Petermarcu** [comment]
Why would each module live in the service folder it supports? Does the springboot community discover the modules through source browsing or is there another listing of modules? What if there was a springboot folder with all the base support modules with a README that linked to the service folder implementations for each service. One of the key things I would like to see happen here is that the spring modules build on top of the Track 2 libraries and get built and tested in the pipelines for each of those services . When those libraries are updated in a way that needs spring module updates, the engineer would be able to do all the work in the context of their service work.

**Comment by JonathanGiles** [comment]
My initial instinct was to have an `/sdk/springboot/` directory structure containing all spring-related code. I've since convinced myself that the better approach is to have the spring libraries co-exist with the service library they are built on top of.

The reason for my initial thinking was discoverability - having all spring code in a single directory meant it was easy to have a single readme and a single place for users to find all available functionality that we offer. As Peter notes, and as was noted in the meeting the other day, this is not the primary way that spring developers discover libraries (they would normally use https://start.spring.io/). Additionally, just because we do decide to co-locate the spring code with the service library does not mean that the readme in /sdk/springboot cannot be maintained with links out to each library, so I don't think we are losing much at all.

The stronger argument to keeping the spring boot code with the service library is to ensure all builds are taking into account the spring code, so that we instantly see build failures and can react immediately.

In short - I agree with the second approach in Mitch's description above.

**Comment by mitchdenny** [comment]
So to close this out I think we are coalescing around the following approach:

1. Were a sprint boot module has a primary relationship with a particular Azure service, that module will be co-located with those modules.
2. Spring boot modules that are not associated with any particular service will exist in ```sdk/springboot``` (including any parents POMs, BOMs etc).

I'd like to propose an intermediate step here. Initially put everything in sdk/springboot and then as part of making the libraries use track 2 versions of the SDK we move them in to the appropriate service directory. This would allow us to get the code itself migrated as is into the springboot directory and start standing up the engineering system elements around it and the slowly we peel off the libraries as appropriate.

**Comment by Petermarcu** [comment]
This plan sounds good to me although I think we should ask ourselves whether we should just be doing the track 2 part initially. Basically, leave track 1 where it is and as things move to track 2 they go into the central repo.

**Comment by mitchdenny** [comment]
We had a follow on conversation with the spring team and initially it will be easier to put them all into ```sdk/spring``` with the intention to break them out when and if we can. There are some relatively tight couplings that will make this complex in the short term.

Additionally they also maintain different versions of the modules on branches to support different versions of Spring Boot/Spring Cloud which adds some complexity about how we bring these two worlds together.

All that said I think that this issue gives us our answer in terms of desired direction. I'm going to close the issue down and when its the right time to break modules out of ```sdk/spring/``` we'll do that.

**Comment by jialindai** [comment]
Sorry for reopen this issue, but I feel we still need to revisit this decision. I'd prefer to keep the Spring modules under sdk/spring:

- It’s easy for Spring developers to visit our repo and browse code
  - We will redirect developers to /sdk/spring from the old repos
- It’s more convenient for daily development as most modules are managed by Shanghai team

**Comment by mitchdenny** [comment]
I guess you've been in the repository for a while now, how have you found integrating changes from various other sdk/[service] folders. Has it been challenging or have you found the inner-loop bearable?

When I opened this thread I indicated my preference of having everything in the spring folder because I assumed that those libraries would co-evolve. But looking forward to when you've fully cut over to track 2 dependencies are they more likely to revolve with the service areas they integrate with or their other peers in the set of spring libraries?

**Comment by Petermarcu** [comment]
Once the work to get to track 2 is all complete, what is the most likely workflow that will happen? Is it changes across many spring modules or is it updating specific modules based on changes in the underlying service libraries?

**Comment by weshaggard** [comment]
Another way to look at what @mitchdenny and @Petermarcu are saying is to think about how these will likely ship. Will they ship as a full set of spring libraries? If so then maybe having them in the sdk/spring folder is the correct answer. Or will they generally need to be shipped when the individual service ships? If so then they should go under the sdk\service folder instead. In general we have been working under the assumption most things should group under the service as opposed to the technology because you will generally need to update all the libraries for a service at the time the service changes, so they are on the shipping schedule matching the service and not the technology they are built on top of.  

Another prospective on why we group under the service folder is that folks coming to our repository are generally going to go into the sdk folder and find the service they want to use and then figure the library of the service they want to use.

**Comment by mitchdenny** [comment]
I'm closing down this issue. I think were we've landed for now is that all spring components stay in the ```sdk/spring``` folder (with the exception of the BOM). The reason for this is that all of the auto configuration logic exists within ```azure-sprint-boot``` and the various starters are just (mostly) codeless wrappers around that logic.

Unless we are going to tear apart ```azure-spring-boot``` into smaller modules and put it in under each service directory. The reasoning is that if we change KeyVault, in order to update the Spring Boot starter we need to update ```azure-spring-boot``` anyway, so it would create a circular dependency between the ```sdk/spring``` directory and the ```sdk/keyvault``` directory.

If folks think I'm wrong on my intuition around closing this down feel free to open it up again :)

**Comment by weshaggard** [comment]
I'm going to re-open this until I can fully understand what is happening here. 

@jialindai based on your comments at https://github.com/Azure/azure-sdk/pull/1155#discussion_r428489545 it would appear that we are supposed to have multiple spring libraries that contain implementation for each service, but that seems to contradict what @mitchdenny is claiming that all the implementation lives in `azure-spring-boot`. As I stated in my comments in the other thread I'm not a spring expert but some things just don't seem to add up for me about how the implementation is factored. It is either in one library, which violates the guidelines in https://github.com/Azure/azure-sdk/pull/1155 or it is split by service in which case I'd still argue that we split the implementations out into the service folders.

**Comment by mitchdenny** [comment]
I got an e-mail from @saragluna on another question related to these discussions. This time it was about ```spring-data-cosmosdb``` and where we should put it. The ```spring-data-cosmosdb``` module is not a Spring Boot module, it is a Spring Data module.

We really have two options for where we place it. The first is in ```sdk/cosmosdb```, and the other is in ```sdk/spring```. Where you place it has more to do with whether it co-evolves with cosmosdb (that is you want to immediately know if changes in cosmos break the Spring module), or whether its more important to be stabilized alongside the rest of the Spring Boot modules.

Regardless of which route we go the module will have to socialize well with the engineering system and the other modules that it co-habitates with.

My gut tells me that this library should co-evolve with cosmosdb, and as part of this migration effort we should update it to be track 2 compliant (for a Spring Data definition of track 2 compliance).

**Comment by mitchdenny** [comment]
Also adding @kushagraThapar to this issue since I've now looped in the ```spring-data-cosmosdb``` thread which he was added to - which is related to [this PR](https://github.com/Azure/azure-sdk-for-java/pull/12287).

**Comment by jialindai** [comment]
Thanks for the discussion and sorry for my late response.

The Spring modules can be roughly divided into following categories, and each of these categories is a separate release unit:

- Azure Spring auto configuration
  - Owned by Azure Spring team
  - The code was in azure-spring-boot repo
  - It includes auto configuration code for services (key vault, aad, cosmos, service bus, storage)
- Cosmos Spring module
  - Owned by cosmos team
  - The code is in spring-data-cosmosdb repo
  - It integrate cosmos SQL API with Spring data API
- Gremlin Spring module
  - Owned by Azure Spring team
  - The code is in spring-data-gremlin repo
  - It integrate cosmos gremlin API with Spring data API
- App config Spring module
  - Owned by App config team
  - The code is in spring-cloud-azure repo
  - It integrate app config with Spring config API
- Messaging Spring module
  - Owned by Azure Spring team
  - The code is in spring-cloud-azure repo
  - It integrate service bus, event hub with Spring Cloud stream API

When considering where to put the code, one category should be considered together.

I'd prefer to keep the categories owned by Azure Spring team under /sdk/spring. Because different categories are released separately, this requires different pipelines under /sdk/spring.

For cosmos Spring module and App config Spring module, it's open to me, I'd like to get inputs from @kushagraThapar and @mrm9084.

**Comment by mitchdenny** [comment]
We don't need separate pipelines to publish separate packages. Whilst it is generally true that we would expect modules co-located under ```sdk/spring``` to ship together at roughly the same time, the engineering system provides the capability to release a subset of the modules (just by approving the package you want to release).

The folder structure isn't really about ownership, its more about co-evolution. So if you want to be in a world where your customers who are using the cosmos spring module can expect that when a new version of cosmos modules are released that there will be a corresponding update to the cosmos-spring module - then it makes sense for them to be co-located, after all I would imagine that cosmos modules would be the major internal dependency for the cosmos-spring module.

Does that make sense?

The benefit of the arrangement is that you can then work on adding unit tests which validate that the updates to cosmos immediately work with your existing spring integration code. If you placed this code under ```sdk/spring``` you wouldn't pick this up until after cosmos had shipped an update and you decided to update the dependency.

Long after you have the opportunity to provide feedback on the Cosmos changes as they come through. The same applies to all other integrations that you've got.

**Comment by kushagraThapar** [comment]
I agree with @mitchdenny and @jialindai  , cosmos spring module (spring-data-cosmosdb) should be placed within sdk/cosmos directory. 

azure-cosmos module is the major and only internal dependency on spring-data-cosmosdb. Every time there is a azure-cosmos release, it is highly likely that spring-data-cosmsodb module also needs to be updated and released. If placing them together also provides the benefit of testing spring-data-cosmosdb module immediately after any changes are made in azure-cosmos module, then that would be the best way going forward. 

Also, azure-spring modules have a release cadence which does not completely inline with spring-data-cosmosdb module's release cadence. spring-data-cosmosdb will be released more frequently based on customer feedback, hotfixes, new changes in azure-cosmos, and so on. 
If spring-data-cosmosdb is placed within azure-spring, then managing those extra releases might just be extra work which can be avoided if spring-data-cosmosdb is to be placed with sdk/cosmos.

**Comment by JimSuplizio** [comment]
@kushagraThapar and I had a chat earlier today and he was able to clarify a few things for me. We then chatted about the best way to proceed.

From the previous posts it seems that we're all in agreement that spring-data-cosmosdb will go into the sdk/cosmos directory. The reason is that there's a tight coupling between azure-cosmos and spring-data-cosmosdb. There is, however, a catch to this: spring-data-cosmosdb is tightly coupled with the v3 azure-cosmos and work will need to get done, by cosmos, for spring-data-cosmosdb to target the v4 version of azure-cosmos. This means that Spring will need to continue targeting the released version of spring-data-cosmos, even after this has been moved into the repository, until this work has been done.

Here's the roadmap of what needs to get done:
**What needs to get done now:**
1. Move the source for spring-data-cosmosdb into sdk/cosmos. (Proabably should rename spring-data-cosmosdb to azure-spring-data-cosmosdb but that could be done afterwards) It should have azure-client-sdk-parent as its parent, it should be added version_client.txt and should also have an entry in the sdk/cosmos/ci.yml file. If it doesn't compile there should be just enough work done to get it compiling/passing in the pipelines. It should continue to have com.microsoft.azure:azure-cosmos as its dependency (andt here's already an entry in external_dependencies, which was previously added by the Spring team) until it has been made v4.
2. As previously mentioned, Spring starters and samples will continue to target the released versions of com.microsoft.azure:spring-data-cosmosdb and com.microsoft.azure:azure-cosmos (again, both already have entries in external_dependencies)

**What will need to get done during spring-data-cosmosdb v4 work:**
1. spring-data-cosmosdb will start using the "current" azure-cosmos version as it's dependency

**What will need to get done after the spring-data-cosmosdb v4 work is considered good enough for Spring to start picking up:**
1. Since the v4 version of spring-data-cosmosdb won't be released, an unreleased_ dependency should be declared in version_client.txt. This dependency is what Spring will use until spring-data-cosmosdb v4 has been shipped.
2. Once the v4 version of spring-data-cosmosdb has been shipped, only then will the unreleased_ dependency be removed from version_client.txt and Spring will then start using the dependency version of Spring.

**Comment by mitchdenny** [comment]
The original intent behind this issue was to build some consensus on how we should approach placing modules in the source tree structure. We used Spring as an example in the Java repository but the intent of the issue is try to get some consistent guidance on how we can structure things within our repositories. I'd like to tie a bow on this issue with some guidance for teams that struggle with this in the future. I'm thinking some general statements like the following:

# Where should your module be located in the source tree?

Most of the ```azure-sdk-for-[language]``` repositories contain an ```sdk/``` folder within the root of the repository. Under that folder you will find many more folders, typically aligned to each specific Azure service - for example ```sdk/cosmos/``` and ```sdk/eventhubs/```.

If your module/library/package is the primary client library that is used to access a service for a particular language then it must be located under the appropriately named ```sdk/[service]``` folder.

In addition to the primary client libraries that must be located under the ```sdk/[service]``` folder, any libraries which interface application frameworks (e.g. Spring Data, ASP.NET, etc) with a particular service should also co-locate their libraries within the ```sdk/[service]``` folder. This is because the primary client library within the service folder represents a internal dependency within the SDK that we would want to validate and test on each incoming PR.

In _limited_ circumstances, libraries for application frameworks **MAY** be located outside of the ```sdk/[service]``` folder structure and instead may occupy a location determined by the name of the application framework (e.g. ```sdk/spring/```). This is only appropriate **IF** the module contains common infrastructure for application framework integration or where all integration with specific services are co-located within a single module. Note that there needs to be a good justification for factoring code this way since it can introduce inefficiencies in the developer inner loop and make some classes of problems more difficult to detect.

**Comment by weshaggard** [comment]
@mitchdenny I'm fine with updating the structure guidance similar to what you called out but can you do that update in our repo layout policy doc https://github.com/Azure/azure-sdk/blob/master/docs/policies/repostructure.md?

However from my prospective this issue is more about trying to decide on whether or not we should allow spring to be an exception to this guidance. From what I've heard so far my stance is still the same which is I feel anything that has a dependency and needs to change with the service should be in the service folder and if it is application framework specific without any real service implementation then it is OK to go under the spring folder. I think of the spring folder much like the core folder in that there are some common pieces needed to work on spring but there are no concrete service implementations included as part of it. It sounds like we have agreement for cosmos and I think we should follow the same model for the other services.

**Comment by mitchdenny** [comment]
Just merged the change into the policy doc. Should we leave this issue open until Azure Spring Boot does the work to break up ```azure-spring-boot```? I'm not sure were all there yet (I agree with you in principle though, I just think the horse has bolted).

**Comment by mrm9084** [comment]
@mitchdenny @jialindai For App Configuration there was hope to unlink our release from those of the other Spring modules. We are actively working on adding new features to the library. In addition we have 4 libraries currently that fall under App Configuration, spring-cloud-azure-appconfiguration-config, spring-cloud-azure-appconfiguration-config-web, spring-cloud-azure-feature-management, and spring-clould-azure-feature-manatement-web.

Also, for our packages their is no dependency on the other Azure Spring packages, and only the two appconfiguration packages have a dependency on from the azure-sdk-for-java, appconfiguration, key vault secrets, and identity.
